gen/src/write.rs

```diff
@@ -373,6 +381,34 @@ fn write_enum(out: &mut OutFile, enm: &Enum) {
     writeln!(out, "}};");
 }
 
+fn check_enum(out: &mut OutFile, enm: &Enum) {
+    let discriminants = enm
+        .variants
+        .iter()
+        .scan(None, |prev_discriminant, variant| {
+            let discriminant = variant
+                .discriminant
+                .unwrap_or_else(|| prev_discriminant.map_or(0, |n| n + 1));
+            *prev_discriminant = Some(discriminant);
+            Some(discriminant)
+        });
+    writeln!(
+        out,
+        "static_assert(sizeof({}) == sizeof(uint32_t));",
+        enm.ident
+    );
+    enm.variants
+        .iter()
+        .zip(discriminants)
+        .for_each(|(variant, discriminant)| {
```

Comment:
I feel like this went a bit overboard with scan/zip/for_each usage. I think this logic would be better suited to a loop.

```rust
let mut prev_discriminant = None;
for variant in &enm.variants {
    let discriminant = variant
        .discriminant
        .unwrap_or_else(|| prev_discriminant.map_or(0, |n| n + 1));
    writeln!(
        out,
        "static_assert({}::{} == {});",
        enm.ident, variant.ident, discriminant,
    );
    prev_discriminant = Some(discriminant);
}
```

Regarding your note in the PR description:

> I'm also considering refactoring the discriminant generation code somewhere now that it's mostly duplicated in three different places.

I agree. I think it would be reasonable to store a precomputed discriminant directly in crate::syntax::Variant (i.e. instead of an Option). Then:

```rust
for variant in &enm.variants {
    writeln!(
        out,
        "static_assert({}::{} == {});",
        enm.ident, variant.ident, variant.discriminant,
    );
}
```

Reply:
The reason I didn't compute the discriminant during parsing in my original patch was to keep it as close to the source as possible.  If you want to change that, we can.

I actually already had a patch refactoring this into a helper method, but it didn't remove the zips so you probably wouldn't like it. :)